go/lib.h

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright GLIDE-for-Redis Project Contributors - SPDX Identifier: Apache-2.0
+ */
+
+#include <stdarg.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+typedef enum RequestErrorType {
+  Unspecified = 0,
+  ExecAbort = 1,
+  Timeout = 2,
+  Disconnect = 3,
+} RequestErrorType;
+
+/**
+ * The connection response.
+ *
+ * It contains either a connection or an error. It is represented as a struct instead of an enum for ease of use in the wrapper language.
+ *
+ * This struct should be freed using `free_connection_response` to avoid memory leaks.
+ */
+typedef struct ConnectionResponse {
+  const void *conn_ptr;
+  const char *connection_error_message;
+} ConnectionResponse;
+
+/**
+ * Success callback that is called when a Redis command succeeds.
+ *
+ * `channel_address` is the address of the Go channel used by the callback to send the error message back to the caller of the command.
+ * `message` is the value returned by the Redis command.
+ */
+typedef void (*SuccessCallback)(uintptr_t channel_address, const char *message);
+
+/**
+ * Failure callback that is called when a Redis command fails.
+ *
+ * `channel_address` is the address of the Go channel used by the callback to send the error message back to the caller of the command.
+ * `error_message` is the error message returned by Redis for the failed command. It should be manually freed after this callback is invoked, otherwise a memory leak will occur.
+ * `error_type` is the type of error returned by glide-core, depending on the `RedisError` returned.
+ */
+typedef void (*FailureCallback)(uintptr_t channel_address,
+                                const char *error_message,
+                                RequestErrorType error_type);
+
+/**
+ * Creates a new client with the given configuration. The success callback needs to copy the given string synchronously, since it will be dropped by Rust once the callback returns. All callbacks should be offloaded to separate threads in order not to exhaust the client's thread pool.
+ *
+ * The returned `ConnectionResponse` should be manually freed by calling `free_connection_response`, otherwise a memory leak will occur. It should be freed whether or not an error occurs.
+ *
+ * `connection_request_bytes` is an array of bytes that will be parsed into a Protobuf `ConnectionRequest` object.
+ * `connection_request_len` is the number of bytes in `connection_request_bytes`.
+ * `success_callback` is the callback that will be called in the case that the Redis command succeeds.
+ * `failure_callback` is the callback that will be called in the case that the Redis command fails.
```

Comment:
nit: maybe a bit clearer:
```suggestion
 * `success_callback` is the callback that will be called when Redis commands executed by the client succeed.
 * `failure_callback` is the callback that will be called when Redis commands executed by the client fail.
```